virtcontainers/clh.go

```diff
@@ -424,22 +432,84 @@ func (clh *cloudHypervisor) hypervisorConfig() HypervisorConfig {
 }
 
 func (clh *cloudHypervisor) resizeMemory(reqMemMB uint32, memoryBlockSizeMB uint32, probe bool) (uint32, memoryDevice, error) {
-	clh.Logger().WithField("function", "resizeMemory").Warn("not supported")
-	return 0, memoryDevice{}, nil
+	if reqMemMB == 0 {
+		return 0, memoryDevice{}, errors.New("Can not resize memory to 0")
+	}
+
+	info, err := clh.vmInfo()
+	if err != nil {
+		return 0, memoryDevice{}, err
+	}
+
+	currentMem := utils.MemUnit(info.Config.Memory.Size) * utils.Byte
+	newMem := utils.MemUnit(reqMemMB) * utils.MiB
+
+	// Early check to verify if boot memory is the same as requested
+	if currentMem == newMem {
+		clh.Logger().WithField("memory", reqMemMB).Debugf("VM already has requested memory")
+		return 0, memoryDevice{}, nil
+	}
+
+	if currentMem > newMem {
+		clh.Logger().Warn("Remove memory is not supported, nothing to do")
+		return 0, memoryDevice{}, nil
+	}
+
+	blockSize := utils.MemUnit(memoryBlockSizeMB) * utils.MiB
+	hotplugSize := (newMem - currentMem).AlignMem(blockSize)
+
+	// Update memory request to increase memory aligned block
+	alignedRequest := currentMem + hotplugSize
+	if newMem != alignedRequest {
+		clh.Logger().WithFields(log.Fields{"request": newMem, "aligned_request": alignedRequest}).Debug("aligning VM memory request")
+		newMem = alignedRequest
+	}
+
+	// Check if memory is the same as requested, a second check is done
+	// to consider the memory request now that is updated to be memory aligned
+	if currentMem == newMem {
+		clh.Logger().WithFields(log.Fields{"current_memory": currentMem, "new_memory": newMem}).Debug("VM already has requested memory(after alignment)")
+		return 0, memoryDevice{}, nil
+	}
```

Comment:
Like the VCPU logic PR, I think this block could be put into a separate function which can then be unit tested. This function would then become very simple:

- Call the new function to work out the resize value (which can be easily unit tested).
- Pass that value to CLH to perform the resize (which cannot be easily unit tested).